packages/react-router/lib/router/router.ts

```diff
@@ -5025,34 +5039,55 @@ async function loadLazyRoute(
         ...mapRouteProperties(routeToUpdate),
         lazy: undefined,
       });
-    });
+    })();
 
     lazyRouteFunctionCache.set(routeToUpdate, lazyRoutePromise);
-    await lazyRoutePromise;
 
-    return;
+    return {
+      lazyRoutePromise,
+      lazyHandlerPromise: lazyRoutePromise,
+    };
   }
 
   let lazyKeys = Object.keys(route.lazy) as Array<keyof typeof route.lazy>;
+  let lazyPropertyPromises: Array<Promise<void>> = [];
+  let lazyHandlerPromise: Promise<void> | undefined = undefined;
 
-  await Promise.all(
-    lazyKeys.map((key) =>
-      loadLazyRouteProperty({
-        key,
-        route,
-        manifest,
-        mapRouteProperties,
-      })
-    )
-  );
+  for (let key of lazyKeys) {
+    let promise = loadLazyRouteProperty({
+      key,
+      route,
+      manifest,
+      mapRouteProperties,
+    });
+    if (promise) {
+      lazyPropertyPromises.push(promise);
+      if (key === type) {
+        lazyHandlerPromise = promise;
+      }
+    }
+  }
+
+  let lazyRoutePromise = Promise.all(lazyPropertyPromises)
+    // Ensure type is Promise<void>, not Promise<void[]>
+    .then(() => {});
+
+  return {
+    lazyRoutePromise,
+    lazyHandlerPromise,
```

Comment:
Not necessarily for this PR but should we be trying to cache/dedupe these as well?